### ISSUE TITLE
Config file support 

### ISSUE THREAD

**Issue by stevenj2019** [BusKill/buskill-app#16]
Config File to manage various settings within the application, 
such as Trigger and Device. 
This will allow for: 
-> Modular Trigger System (expanding to a DIY Trigger oppurunity) 
-> Preferred Device (i.e. if you have a device you routinely removed and would like BusKill to ignore its event) 
-> a "First Start" sequence to allow for first time configuration

Once Configuration Files are an option within BusKill we can perform further enhancements such as: 
-> update frequency
-> Trigger EcoSystem

**Comment by mrx23dot** [comment]
Just have a config.txt, look for 
pre_shutdown = 'os calls'  if exists execute

`
import configparser

def configLoad(initConf):
  """ creates file if doesn't exist """
  global CONFIG_FILE
  readDict = dict()
  config = configparser.ConfigParser()
  try:
    config.read_file(open(CONFIG_FILE))
    readDict = dict(config[configparser.DEFAULTSECT])
  except Exception as e:
    print(' exception [config load]',e) 
  return dict(initConf, **readDict)

`

**Comment by maltfield** [comment]
TODO: Once this is done, please move `set_trigger()` to automatically run on __init__ of the BusKill object, based on the default trigger that's defined in the config file.

**Comment by maltfield** [comment]
I started working on this in the Trigger Selector ticket. See here for more info:

 * https://github.com/BusKill/buskill-app/issues/14#issuecomment-1299341411

**Comment by maltfield** [comment]
It appears that I'm already using this kivy Config module! I set two options to change the following behaviours:

1. `exit_on_escape = 0` to prevent the default behaviour where if the user presses the `esc` key on their keyboard, the app closes https://github.com/BusKill/buskill-app/commit/56dc42b9a9fe8b6ddce9dce19731ceff56e720fa
2. disable `multitouch` emulation to prevent the default behaviour where a strange orange circle would appear if the user right-clicked anywhere on the app (it's multitouch emulation) https://github.com/BusKill/buskill-app/commit/3c8625dfcb673dfb43bf5c3308b5b0a7ade9f4e3

Interestingly these settings didn't persist in the kivy config file. My guess is that's because I'm only applying them at runtime at the start of `buskill_gui.py` every time. I didn't even know there was a config file that these options mapped-to for persistence. I think I'd just need to call `write()` to make them persist to the file.

 * https://kivy.org/doc/stable/api-kivy.config.html#kivy.config.ConfigParser.write

```
user@buskill:~$ cat .kivy/config.ini 
[kivy]
keyboard_repeat_delay = 300
keyboard_repeat_rate = 30
log_dir = logs
log_enable = 1
log_level = info
log_name = kivy_%y-%m-%d_%_.txt
window_icon = 
keyboard_mode = 
keyboard_layout = qwerty
desktop = 1
exit_on_escape = 1
pause_on_minimize = 0
kivy_clock = default
default_font = ['Roboto', 'data/fonts/Roboto-Regular.ttf', 'data/fonts/Roboto-Italic.ttf', 'data/fonts/Roboto-Bold.ttf', 'data/fonts/Roboto-BoldItalic.ttf']
log_maxfiles = 100
window_shape = data/images/defaultshape.png
config_version = 21

[graphics]
display = -1
fullscreen = 0
height = 600
left = 0
maxfps = 60
multisamples = 2
position = auto
rotation = 0
show_cursor = 0
top = 0
width = 800
resizable = 1
borderless = 0
window_state = visible
minimum_width = 0
minimum_height = 0
min_state_time = .035
allow_screensaver = 1
shaped = 0
vsync = 
verify_gl_main_thread = 1
custom_titlebar = 0
custom_titlebar_border = 5

[input]
mouse = mouse
%(name)s = probesysfs

[postproc]
double_tap_distance = 20
double_tap_time = 250
ignore = []
jitter_distance = 0
jitter_ignore_devices = mouse,mactouch,
retain_distance = 50
retain_time = 0
triple_tap_distance = 20
triple_tap_time = 375

[widgets]
scroll_timeout = 250
scroll_distance = 20
scroll_friction = 1.
scroll_stoptime = 300
scroll_moves = 5

[modules]

[network]
useragent = curl

user@buskill:~$
```

**Comment by maltfield** [comment]
In the above commits, I was successfully able to combine the kivy and buskill files into one ini config file, but unfortunately I cannot prevent kivy from creating the `~/.kivy/` dir because the logic for determining the buskill `DATA_DIR` is in the BusKill object's `__init__()` function. But the kivy dir is setup earlier, before that object is instantiated.

Kivy will allow you to specify where its data dir is with the environment variable `KIVY_HOME`.

 * https://kivy.org/doc/stable/guide/config.html

So the only way to catch this early enough would be to move the logic for determining DATA_DIR to `main.py`, and then set `KIVY_HOME` and `BUSKILL_DATA_DIR` at that time. This would have the added benefit of being able to put out buskill log files there, too.

**Comment by maltfield** [comment]
Today I successfully moved the instantiation of the buskill object into `main.py`, so it happens before the GUI or CLI scripts are called. This allowed me to set the `KIVY_HOME` environment variable to the buskill `DATA_DIR` directory before the kivy app is ever created, and so now the kivy `config.ini` directory (and all other kivy files) are being created in our main buskill `DATA_DIR` as desired

**Comment by maltfield** [comment]
Today I successfully got the kivy built-in settings widget to be loaded in the Settings screen of the BusKill app.

TODO: figure-out how to limit the settings items that are enumerated to only those in the `[buskill]` section of the `config.ini` file

**Comment by maltfield** [comment]
ok, using the built-in settings widget and a new json file, I was successfully able to limit the config items appearing in the Settings Screen to only the "trigger" key/value, as defined in a json template (which is actually distinct from where the *value* is stored-to--that's still always the .ini file)

Though this is very robust, I still need to extend the Settings widget because:

1. When the user chooses between `lock-screen` and `soft-shutdown`, there is no subtitle between the text that explains what each option does
2. On the main "Settings" screen, there *is* a subtitle, but it doesn't change with each option. It should say what the currently-selected option does, not generically what the `trigger` option means.

**Comment by maltfield** [comment]
See also https://stackoverflow.com/questions/74452839/nameerror-class-is-not-defined-when-overriding-a-class-in-python?noredirect=1#comment131432280_74452839

**Comment by maltfield** [comment]
Today I've been trying to override the methods of the Kivy Settings widget so that I can customize it a bit, but it's a gnarly mess.

I spent some time trying to understand what the heck is in this `s` variable's `SettingsWithNoMenu()` class instance. For that, I hacked this code

```
                        print( str( s.children ) )
                        print( str( s.children[0].children ) )
                        print( str( s.children[0].children[0].children ) )
                        print( str( s.children[0].children[0].children[0].children ) )
                        print( "SettingOptions: " +str( s.children[0].children[0].children[0].children[0].children ) )
                        print( "\t" + str( s.children[0].children[0].children[0].children[0].children[0].children ) )
                        print( "\tBoxLayout: " + str( s.children[0].children[0].children[0].children[0].children[0].children[0].children ) )
                        print( "\t\t" + str( s.children[0].children[0].children[0].children[0].children[0].children[0].children[0].children ) )
                        print( "\tLabel: " + str( s.children[0].children[0].children[0].children[0].children[0].children[1].children ) )
                        print( "Label: " +str( s.children[0].children[0].children[0].children[1].children ) )
```

There's probably a better, built-in pythonic way to see this, but I just manually added these over-and-over until I was able to see the whole tree. Here's the output:

```
<kivy.uix.settings.InterfaceWithNoMenu object at 0x717f55169600>]
[<kivy.uix.gridlayout.GridLayout object at 0x717f55172ad0>]
[<kivy.uix.settings.SettingsPanel object at 0x717f551077c0>]
[<kivy.uix.settings.SettingOptions object at 0x717f5511abb0>, <kivy.uix.label.Label object at 0x717f5510a4b0>]
SettingOptions: [<kivy.uix.boxlayout.BoxLayout object at 0x717f551229f0>]
	[<kivy.uix.boxlayout.BoxLayout object at 0x717f55135980>, <kivy.uix.label.Label object at 0x717f5512ba60>]
	BoxLayout: [<kivy.uix.label.Label object at 0x717f5513bbb0>]
		[]
	Label: []
Label: []
```

Here's the human-readable version: what we have is

The `s` variable itself is a `BusKillSettingsWithNoMenu`. That object is just my custom object that overrides the [kivy.uix.settings.SettingsWithNoMenu](https://github.com/kivy/kivy/blob/2efdc2646147e5604767b8a5c6a43e5808455bf2/kivy/uix/settings.py#L1140).

The `BusKillSettingsWithNoMenu` object has one child = a `kivy.uix.settings.InterfaceWithNoMenu` object

The `InterfaceWithNoMenu` object has one child = a kivy `GridLayout`

This `GridLayout` has one child = a `kivy.uix.settings.SettingsPanel` object

The `SettingsPanel` has two children: a `kivy.uix.settings.SettingOptions` object and a kivy `Label` object.

The `Label` has no children.

The `SettingOptions` object has one child = a kivy `BoxLayout`

The `BoxLayout` two children: another `BoxLayout` and a `Label`.

The `Label` has no children.

This nested `BoxLayout` has one child = another kivy `Label`.

The `Label` has no children.

**Comment by maltfield** [comment]
From above, I'd guess that the `Label` child of the `SettingsPanel` is the title that I wanted to remove

`s.interface` has to useful attributes so I can cut straight through this tree to find the `SettingsPanel`. Currently we only have one panel, but you can get a list of the panels and also the current panel as follows:

```
                        print( "s.interface.panels:|" +str(s.interface.panels)+ "|" )
                        print( "s.interface.current_panel:|" +str(s.interface.current_panel)+ "|" )
```

Example execution:

```
s.interface:|<kivy.uix.settings.InterfaceWithNoMenu object at 0x7fe9c5eff600>|
s.interface.panels:|{602: <kivy.uix.settings.SettingsPanel object at 0x7fe9c5e9d7c0>}|
s.interface.current_panel:|<kivy.uix.settings.SettingsPanel object at 0x7fe9c5e9d7c0>|
s.interface.children:|[<kivy.uix.gridlayout.GridLayout object at 0x7fe9c5f07ad0>]|
s.interface.walk():|[<kivy.uix.settings.InterfaceWithNoMenu object at 0x7fe9c5eff600>, <kivy.uix.gridlayout.GridLayout object at 0x7fe9c5f07ad0>, <kivy.uix.settings.SettingsPanel object at 0x7fe9c5e9d7c0>, <kivy.uix.label.Label object at 0x7fe9c5ea04b0>, <kivy.uix.settings.SettingOptions object at 0x7fe9c5eb0bb0>, <kivy.uix.boxlayout.BoxLayout object at 0x7fe9c5eb89f0>, <kivy.uix.label.Label object at 0x7fe9c5ec0a60>, <kivy.uix.boxlayout.BoxLayout object at 0x7fe9c5ecb980>, <kivy.uix.label.Label object at 0x7fe9c5ed1bb0>]|
dir(s.interface):|['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__events__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__metaclass__', '__module__', '__ne__', '__new__', '__proxy_getter', '__proxy_setter', '__pyx_vtable__', '__reduce__', '__reduce_ex__', '__repr__', '__self__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '_apply_transform', '_bar_color', '_bind_inactive_bar_color', '_bind_inactive_bar_color_ev', '_change_bar_color', '_change_touch_mode', '_context', '_disabled_count', '_disabled_value', '_do_touch_up', '_effect_x_start_width', '_effect_y_start_height', '_get_do_scroll', '_get_hbar', '_get_uid', '_get_vbar', '_handle_x_pos', '_handle_x_size', '_handle_y_pos', '_handle_y_size', '_kwargs_applied_init', '_proxy_ref', '_set_do_scroll', '_set_viewport_size', '_touch', '_touch_in_handle', '_trigger_update_from_scroll', '_update_effect_bounds', '_update_effect_bounds_ev', '_update_effect_widget', '_update_effect_x', '_update_effect_x_bounds', '_update_effect_y', '_update_effect_y_bounds', '_viewport', '_walk', '_walk_reverse', 'add_panel', 'add_widget', 'apply_class_lang_rules', 'apply_property', 'bar_color', 'bar_inactive_color', 'bar_margin', 'bar_pos', 'bar_pos_x', 'bar_pos_y', 'bar_width', 'bind', 'canvas', 'canvas_viewport', 'center', 'center_x', 'center_y', 'children', 'clear_widgets', 'cls', 'collide_point', 'collide_widget', 'container', 'convert_distance_to_scroll', 'create_property', 'current_panel', 'current_uid', 'dec_disabled', 'disabled', 'dispatch', 'dispatch_children', 'dispatch_generic', 'do_scroll', 'do_scroll_x', 'do_scroll_y', 'effect_cls', 'effect_x', 'effect_y', 'events', 'export_as_image', 'export_to_png', 'fbind', 'funbind', 'g_translate', 'get_center_x', 'get_center_y', 'get_disabled', 'get_parent_window', 'get_property_observers', 'get_right', 'get_root_window', 'get_top', 'get_window_matrix', 'getter', 'hbar', 'height', 'id', 'ids', 'inc_disabled', 'is_event_type', 'on__viewport', 'on_current_uid', 'on_effect_cls', 'on_effect_x', 'on_effect_y', 'on_kv_post', 'on_opacity', 'on_scroll_move', 'on_scroll_start', 'on_scroll_stop', 'on_touch_down', 'on_touch_move', 'on_touch_up', 'opacity', 'panels', 'parent', 'pos', 'pos_hint', 'properties', 'property', 'proxy_ref', 'register_event_type', 'remove_widget', 'right', 'scroll_distance', 'scroll_timeout', 'scroll_to', 'scroll_type', 'scroll_wheel_distance', 'scroll_x', 'scroll_y', 'set_center_x', 'set_center_y', 'set_disabled', 'set_right', 'set_top', 'setter', 'simulate_touch_down', 'size', 'size_hint', 'size_hint_max', 'size_hint_max_x', 'size_hint_max_y', 'size_hint_min', 'size_hint_min_x', 'size_hint_min_y', 'size_hint_x', 'size_hint_y', 'smooth_scroll_end', 'to_local', 'to_parent', 'to_widget', 'to_window', 'top', 'uid', 'unbind', 'unbind_uid', 'unregister_event_types', 'update_from_scroll', 'vbar', 'viewport_size', 'walk', 'walk_reverse', 'width', 'x', 'y']|
```

**Comment by maltfield** [comment]
If I try to set it to the empty string, it works -- but there is an empty label there taking up space. If I try to set it to `None`

```
                        s.interface.current_panel.title = None
```

..then I get an error

   File "/home/user/sandbox/buskill-app/src/buskill_gui.py", line 613, in on_pre_enter
     s.interface.current_panel.title = None
   File "kivy/properties.pyx", line 497, in kivy.properties.Property.__set__
   File "kivy/properties.pyx", line 541, in kivy.properties.Property.set
   File "kivy/properties.pyx", line 532, in kivy.properties.Property.set
   File "kivy/properties.pyx", line 695, in kivy.properties.StringProperty.check
   File "kivy/properties.pyx", line 570, in kivy.properties.Property.check
 ValueError: None is not allowed for SettingsPanel.title
```
```

**Comment by maltfield** [comment]
SettingOptions is defined, but it leaves much to wonder how it works

```<SettingOptions>:
    Label:
        text: root.value or ''
        pos: root.pos
        font_size: '15sp'
```

...but when you check the code, you see that `SettingOptions` is a child of `SettingItem`

https://github.com/kivy/kivy/blob/2efdc2646147e5604767b8a5c6a43e5808455bf2/kivy/uix/settings.py#L612

```
<SettingItem>:
    size_hint: .25, None
    height: labellayout.texture_size[1] + dp(10)
    content: content
    canvas:
        Color:
            rgba: 47 / 255., 167 / 255., 212 / 255., self.selected_alpha
        Rectangle:
            pos: self.x, self.y + 1
            size: self.size
        Color:
            rgb: .2, .2, .2
        Rectangle:
            pos: self.x, self.y - 2
            size: self.width, 1

    BoxLayout:
        pos: root.pos

        Label:
            size_hint_x: .66
            id: labellayout
            markup: True
            text: u'{0}\n[size=13sp][color=999999]{1}[/color][/size]'.format(root.title or '', root.desc or '')
            font_size: '15sp'
            text_size: self.width - 32, None

        BoxLayout:
            id: content
            size_hint_x: .33
```

```
```

**Comment by maltfield** [comment]
See also https://stackoverflow.com/questions/74467933/show-to-override-built-in-kivy-widgets-and-their-kv-language-layouts

**Comment by maltfield** [comment]
I finally figured out why overriding a kivy widget's kv-language only added-to (instead of replaced) the existing widget's layout.

 * https://kivy.org/doc/stable/api-kivy.lang.html#redefining-a-widget-s-style

As the above documentation describes, you need to change

```
<SomeWidget>
```

with this (you add a dash before it to clear it)

```
<-SomeWidget>
```

**Comment by maltfield** [comment]
See also https://stackoverflow.com/questions/74480089/equivalent-of-browser-debuggers-box-model-inspector-in-kivy

**Comment by maltfield** [comment]
See also https://stackoverflow.com/questions/74482717/why-is-declaring-size-x-and-size-y-separately-different-from-delcaring-both

**Comment by maltfield** [comment]
Yesterday I finished replacing the popup when choosing the value for a given key that's an option (equivalent of a radio button in html).

Instead of doing this in a very restrictive and poor-UX modal popup, we dynamically create a new screen where the user can have more information about each option in a scrollable view without it being cluttered in the modal. This includes:

1. Each row will have a radio button (selected or unselected), a label that has a title for each option and a short description of what that option does, and an optional "gear" icon to further configure that option
2. A "help" ActionItem at the top that, when clicked, presents a modal with a very detailed explanation of what this option does

The layout of rows of options will be similar to the layout of rows of settings on the previous screen, which is

TODO

**Comment by maltfield** [comment]
Yesterday I got the changes made in the GUI's Settings Window to actually write the changes to the now-combined buskill/kivy .ini file

**Comment by maltfield** [comment]
Today I finished with the UI for displaying the help message when a "help" button icon is pressed in the BusKillSettingsComplexOptionsScreen's ActionBar

TODO: implement the "reset" button to restore the defaults for all settings (after confirmation) on the main Settings screen

**Comment by maltfield** [comment]
Today I finished with the logic to actually update the runtime BusKill object's `trigger` after the user changed it when leaving the Settings Screen, and to prompt them to disarm & arm again if they changed the trigger from what it is currently armed with.

I also noticed that nav drawer is open when exiting the Settings Screen.

TODO: fix nav-drawer-open bug
TODO: implement logic for the "rest" button to restore all buskill settings to their defaults

**Comment by maltfield** [comment]
Note to self: if you see this error

```
 Traceback (most recent call last):
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/configparser.py", line 788, in get
     value = d[option]
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/collections/__init__.py", line 916, in __getitem__
     return self.__missing__(key)            # support subclasses that define __missing__
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/collections/__init__.py", line 908, in __missing__
     raise KeyError(key)
 KeyError: 'drive'
 
 During handling of the above exception, another exception occurred:
 
 Traceback (most recent call last):
   File "src/main.py", line 146, in <module>
     BusKillApp( bk ).run()
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/app.py", line 855, in run
     runTouchApp()
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/base.py", line 504, in runTouchApp
     EventLoop.window.mainloop()
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/core/window/window_sdl2.py", line 747, in mainloop
     self._mainloop()
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/core/window/window_sdl2.py", line 479, in _mainloop
     EventLoop.idle()
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/base.py", line 342, in idle
     self.dispatch_input()
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/base.py", line 327, in dispatch_input
     post_dispatch_input(*pop(0))
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/base.py", line 293, in post_dispatch_input
     wid.dispatch('on_touch_up', me)
   File "kivy/_event.pyx", line 707, in kivy._event.EventDispatcher.dispatch
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/uix/behaviors/button.py", line 179, in on_touch_up
     self.dispatch('on_release')
   File "kivy/_event.pyx", line 703, in kivy._event.EventDispatcher.dispatch
   File "kivy/_event.pyx", line 1214, in kivy._event.EventObservers.dispatch
   File "kivy/_event.pyx", line 1098, in kivy._event.EventObservers._dispatch
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/lang/builder.py", line 64, in custom_callback
     exec(__kvlang__.co_value, idmap)
   File "/home/user/sandbox/buskill-app/src/buskill.kv", line 108, in <module>
     root.manager.current = 'settings'
   File "kivy/properties.pyx", line 497, in kivy.properties.Property.__set__
   File "kivy/properties.pyx", line 544, in kivy.properties.Property.set
   File "kivy/properties.pyx", line 599, in kivy.properties.Property.dispatch
   File "kivy/_event.pyx", line 1214, in kivy._event.EventObservers.dispatch
   File "kivy/_event.pyx", line 1120, in kivy._event.EventObservers._dispatch
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/uix/screenmanager.py", line 1049, in on_current
     self.transition.start(self)
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/uix/screenmanager.py", line 377, in start
     self.screen_in.dispatch('on_pre_enter')
   File "kivy/_event.pyx", line 707, in kivy._event.EventDispatcher.dispatch
   File "/home/user/sandbox/buskill-app/src/buskill_gui.py", line 1032, in on_pre_enter
     s.add_json_panel( 'buskill', Config, os.path.join(self.bk.SRC_DIR, 'packages', 'buskill', 'settings_buskill.json') )
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/uix/settings.py", line 975, in add_json_panel
     panel = self.create_json_panel(title, config, filename, data)
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/uix/settings.py", line 1015, in create_json_panel
     instance = cls(panel=panel, **str_settings)
   File "/home/user/sandbox/buskill-app/src/buskill_gui.py", line 751, in __init__
     super(BusKillSettingComplexOptions, self).__init__(**kwargs)
   File "/home/user/sandbox/buskill-app/src/buskill_gui.py", line 680, in __init__
     super(BusKillSettingItem, self).__init__(**kwargs)
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/uix/settings.py", line 289, in __init__
     self.value = self.panel.get_value(self.section, self.key)
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/uix/settings.py", line 654, in get_value
     return config.get(section, key)
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/site-packages/kivy/config.py", line 501, in get
     value = PythonConfigParser.get(self, section, option, **kwargs)
   File "/tmp/kivy_appdir/opt/python3.7/lib/python3.7/configparser.py", line 791, in get
     raise NoOptionError(option, section)
 configparser.NoOptionError: No option 'drive' in section: 'buskill'
 [INFO/MainProcess] process shutting down
 [DEBUG/MainProcess] running all "atexit" finalizers with priority >= 0
 [DEBUG/MainProcess] running the remaining "atexit" finalizers
```

..then it means you need to add a key/value pair to the defaults defined by `Config.setdefaults()` in `BusKillApp()`'s `build_config()` in `buskill_gui.py`

```
        def build_config(self, config):

                Config.read( self.bk.CONF_FILE )
                Config.setdefaults('buskill', {
                 'trigger': 'lock-screen',
                })      
                Config.set('kivy', 'exit_on_escape', '0')
                Config.set('input', 'mouse', 'mouse,multitouch_on_demand')
                Config.write()
```

**Comment by maltfield** [comment]
I fixed a logic bug where I was attempting to re-arm whenever leaving the Settings screen -- but it really should only happen when leaving the Settings Screen *and* going back to the Main Screen. It should *not* attempt to re-arm when leaving the Settings Screen *and* going into a ComplexOptions Screen.

TODO: fix a lot of bugs where the value is correct in the config file after reset, but the value is *not* correct on the Settings Screen or the wrong option is selected in the ComplexOption Screen

**Comment by maltfield** [comment]
I'm a bit confused as to why my `BusKillOptionItem()` instances have the values set to `title ` and the `value` is hardcoded to `replaceme`. Currently the only way to get the title is by going up the tree to the screen and getting its name.

For example for the "trigger" option, I'd think that the title should be `trigger` and the `value` either `lock-screen` or `soft-shutdown`. But the `title` is currently `lock-screen` or `soft-shutdown` and the `value` is `replaceme`.

I'd like to change this, but it may break things and take a while to fix :(

**Comment by maltfield** [comment]
I fixed the `title` / `value` thing above, and I also finished implementing the "Reset to Defaults" button functionality. I'm not focusing on two bugs:

1. The nav bar drawer is open if you leave the Settings screen. This only happens if you went to a ComplexOption sub-screen first
2. I'm getting these weird warning messages about having multiple copies of the ComplexOption sub-screens

```
[WARNING] [Multiple screens named "setting_trigger"] [<Screen name='setting_trigger'>, <Screen name='setting_trigger'>, <Screen name='setting_trigger'>]
```

The above message does not show-up the first time I click on the "trigger" option on the Settings Screen. It does show-up if I go back and then click on "trigger" again. The thrid time I click on it, I get this

```
[WARNING] [Multiple screens named "setting_trigger"] [<Screen name='setting_trigger'>, <Screen name='setting_trigger'>, <Screen name='setting_trigger'>, <Screen name='setting_trigger'>, <Screen name='setting_trigger'>]
```

And the fourth

```
[WARNING] [Multiple screens named "setting_trigger"] [<Screen name='setting_trigger'>, <Screen name='setting_trigger'>, <Screen name='setting_trigger'>, <Screen name='setting_trigger'>, <Screen name='setting_trigger'>, <Screen name='setting_trigger'>, <Screen name='setting_trigger'>]
```

**Comment by maltfield** [comment]
ok, I fixed the infinite screen creation

**Comment by maltfield** [comment]
I also fixed the navbar drawer staying open.

As I cleanup the code, I went ahead and added some test options for all the other kivy types

 * https://kivy.org/doc/stable/api-kivy.uix.settings.html#create-a-panel-from-json

While these new options generally "just work" -- the reset button does *not* work for them.

TODO: see if clearing the widgets from the Settings screen and re-running the __init__() (as we were doing before) now works for those. I think maybe the "multiple screens" bug may have caused that to not work? Otherwise I'll just have to iterate into each of those settings and update them to match the config file, not too hard.

**Comment by maltfield** [comment]
I fixed the reset button to reset the built-in kivy types too.

For future reference (when we will add more options that my be of the built-in kivy types), here's what that looks like:

https://github.com/BusKill/buskill-app/blob/fbfa7fa3c305ca6873458804d4ace7d7bc516806/src/packages/buskill/settings_buskill.json#L25-L60

https://github.com/BusKill/buskill-app/blob/fbfa7fa3c305ca6873458804d4ace7d7bc516806/src/buskill_gui.py#L1163-L1172

**Comment by maltfield** [comment]
Today I did a lot more cleanup of the `buskill_gui.py` file and the `buskill.kv` file. I commented-out all the canvas bits for awkward colors that I used to see the edge of labels, and I removed all the other test SettingItems and the 'hard-shutdown' OptionItem (as we haven't implemented that trigger yet -- I was just testing the the GUI for the future).

At this point, I think this feature might be complete? I think I should just test the build on all three platforms. If all is well, then maybe I can merge it into dev!
